### PR TITLE
Make --enable-debug to pick better options

### PR DESCRIPTION
**Pull request by practicalswift**
Cherry-picked (and rebased) 94189645e67f364c4445d62e2b00c282d885cbbf from the "up for grabs" PR: "[build] Make --enable-debug pick better options" (#12695).

See previous review in #12695.

### CODE REVIEW

**Comment by practicalswift** [comment]
One small thing that needs to be fixed before merge is that `CPPFLAGS`, `CXXFLAGS` and `LDFLAGS` gets stuffed with redundant spaces when printing:

```
$ ./configure --with-incompatible-bdb
…
  CC            = gcc
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = g++ -std=c++11
  CXXFLAGS      =   -Wstack-protector -fstack-protector-all   -g -O2 -Wall -Wextra -Wformat -Wvla -Wformat-security -Wno-unused-parameter
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie
  ARFLAGS       = cr
```

I'll find a way to fix that.

**Comment by practicalswift** [comment]
Added a commit fixing the annoying extra whitespace:

```
$ ./configure
…

  CC            = gcc
  CFLAGS        = -g -O2
  CPPFLAGS      = -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = g++ -std=c++11
  CXXFLAGS      = -Wstack-protector -fstack-protector-all -g -O2 -Wall -Wextra -Wformat -Wvla -Wformat-security -Wno-unused-parameter
  LDFLAGS       = -pthread -Wl,-z,relro -Wl,-z,now -pie
  ARFLAGS       = cr
```

**Comment by laanwj** [comment]
re-utACK
Ping @theuni - he proposed some changes in #12695, so would like to know if he's ok with the current state.

**Comment by ryanofsky** [comment]
utACK 8a0ad1a2333b329d6eb8d34a25277181ee675f23. The first commit is just rebased and has no changes since my previous review in the other PR. The second commit is new and looks ok, though I think the echo_joined function is overkill. I think you should just drop quotes around the variables and let the shell collapse spaces:

```sh
echo "  CXXFLAGS      =" $DEBUG_CXXFLAGS $HARDENED_CXXFLAGS $ERROR_CXXFLAGS $CXXFLAGS
```

**Comment by theuni** [comment]
Agree with @ryanofsky about echo_joined. I'd prefer to just drop the second commit.
utACK otherwise.

**Comment by practicalswift** [comment]
@eklitzke Good idea about fixing the whitespace with `AX_APPEND_COMPILE_FLAGS`. Could you help with a diff for that?

**Comment by theuni** [comment]
@practicalswift That's worth handling in a separate PR. It will be a pretty huge diff.

**Comment by practicalswift** [comment]
@theuni Oh, I see.

@eklitzke @theuni Assuming `AX_APPEND_COMPILE_FLAGS` is fixed in a follow-up PR do you have time re-review this PR as-is and give your ACK/utACK/NACK? :-)

**Comment by fanquake** [comment]
@practicalswift From reading the above discussion, and on the [original PR](https://github.com/bitcoin/bitcoin/pull/12695#issuecomment-377032058), you still need to drop the second commit. Then this can be reviewed, and a follow-up PR can introduce the `AX_APPEND_FLAG / AX_APPEND_COMPILE_FLAGS` changes.

**Comment by practicalswift** [comment]
@fanquake Updated! Please re-review! :-)

**Comment by laanwj** [comment]
re-utACK 9e49db2426cd4508d9dd95c7461e107bab7fd7de

File: configure.ac [hunk 1]
```diff
@@ -243,14 +243,22 @@ AC_LANG_PUSH([C++])
...
-    CPPFLAGS="$CPPFLAGS -DDEBUG -DDEBUG_LOCKORDER"
-    if test "x$GCC" = xyes; then
-        CFLAGS="$CFLAGS -g3 -O0"
-    fi
+  # Prefer -Og, fall back to -O0 if that is unavailable.
+  AX_CHECK_COMPILE_FLAG(
+    [-Og],
+    [[DEBUG_CXXFLAGS="$DEBUG_CXXFLAGS -Og"]],
+    [AX_CHECK_COMPILE_FLAG([-O0],[[DEBUG_CXXFLAGS="$DEBUG_CXXFLAGS -O0"]],,[[$CXXFLAG_WERROR]])],
+    [[$CXXFLAG_WERROR]])
```
Comment (luke-jr) [inline configure.ac]: What if `-Og` *and* `-O0` are unavailable? (Both of these may compromise spectre mitigations, so I have my x86 compiler not support them at all)

Comment (practicalswift) [reply configure.ac]: Good question! @eklitzke, what do you say? :-)

Comment (theuni) [reply configure.ac]: I'd prefer making this -O1. Completely un-optimized code barely resembles anything that ever gets run, and macro/inline expansions make it hard to deal with anyway.

**Pull request merged**